### DISCUSSION TITLE
Determining differences between builds

### DISCUSSION THREAD

**Discussion by nosjojo** [pyinstaller/pyinstaller#7326] [Help] [answered]
I've got a build that uses the python package [pyaardvark](https://github.com/kontron/python-aardvark), and one of the recent versions is failing during execution. The failure itself is an error code coming from the vendor DLL, so there aren't any pyinstaller/runtime errors.

If I run the code in the python interpreter, it works fine with no issues. It's not until I build and run the executable that the error appears. I assume it's something related to a dependency that's incorrect/missing.

Reviewing the source code differences between the working and non-working versions though, the only real difference is the update to the newest version of the aardvark.pyd file that is distributed by the manufacturer. So that's probably the 'source' of the issue. But what I'm stumped on is trying to fix my build to compensate for it.

I ran both the working and non-working builds with the --log-level set to TRACE and did some rough comparisons (tossed out duplicate lines, sorted both, compared where they differed in behavior). The only differences I saw with this was the addition of the OSXARM pyd file (which should be ignored here), and a few extra lines in the working version:

1. Adding VCRUNTIME140.dll dependency of python.exe from `<env path>\VCRUNTIME140.dll`
    - This dll and dependency is listed in both, the working one just detects it for both python.exe and python39.dll
2. Skipping ADVAPI32.dll dependency of aardvark.pyd
    - This one is listed several times, the non-working version doesn't have any ADVAPI32.DLL instances for aardvark.pyd, but both list instances of 'skipping ADVAPI32.dll` for various DLLs.

Both versions have identical warnings for missing modules (the newest has an entry for the skipped OSXARM dll, only difference).

Is there any other sort of debug/analysis I can do to figure out what might be missing? I reached out to the company to see if they had any idea, but since this is related to the build and not the normal usage, I don't expect much there.

**Comment by rokm** [comment, marked as answer]
If you are referring to the 
```
Unable to bind to Python API
API supports Python 2.5+ and 3.2+
```
then the problem is missing `python3.dll`. 

Instead of being link-time dependency, the extension is probably trying to load it dynamically at runtime, so our dependency analysis does not pick it up automatically.

---

You can manually ensure its collection via spec file, by adding 

```python
from PyInstaller.depend import bindepend

binaries = []
python3_dll = bindepend.findSystemLibrary('python3.dll')
if python3_dll:
    binaries += [(python3_dll, '.')]
```

at the top of the .spec file, and then passing the `binaries` to the `Analysis`:

```
a = Analysis(
    ...
    binaries=binaries,
    ...
```

And then rebuild your program by running PyInstaller against the .spec file instead of the .py file.

Or, you can create a custom hook for `pyaardvark` in directory called `extra-hooks`
```
# extra-hooks/hook-pyaardvark.py
from PyInstaller.depend import bindepend

binaries = []
python3_dll = bindepend.findSystemLibrary('python3.dll')
if python3_dll:
    binaries += [(python3_dll, '.')]
``` 
and run PyInstaller with added `--additional-hooks-dir extra-hooks` argument.

**Reply by nosjojo** [reply]
Thanks so much, that worked perfectly. I'll have to add a note to pyaardvark so this doesn't burn others in the future.

**Reply by rqx110** [reply]
enhancement(get python3.dll from same environment):
```py
import os
from PyInstaller.depend import bindepend

binaries = []
python_lib = bindepend.get_python_library_path()
python3_dll = bindepend.resolve_library_path(
    'python3.dll', 
    search_paths=[os.path.dirname(python_lib)],
)
if python3_dll:
    binaries += [(python3_dll, '.')]
```
